staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"net/netip"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// IPAddr provides a CEL function library extension of IP address parsing functions.
+//
+// ipAddr
+//
+// Converts a string to an IP address or results in an error if the string is not a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+//
+//	ipAddr(<string>) <IPAddr>
+//
+// Examples:
+//
+//	ipAddr('127.0.0.1') // returns an IPv4 address
+//	ipAddr('::1') // returns an IPv6 address
+//	ipAddr('127.0.0.256') // error
+//	ipAddr(':::1') // error
+//
+// isIPAddr
+//
+// Returns true if a string is a valid IP address. The IP address must be an IPv4 or IPv6 address.
+//
+//	isIPAddr(<string>) <bool>
+//
+// Examples:
+//
+//	isIPAddr('127.0.0.1') // returns true
+//	isIPAddr('::1') // returns true
+//	isIPAddr('127.0.0.256') // returns false
+//	isIPAddr(':::1') // returns false
+//
+// isIPv4 / isIPv6 / isInCIDR
+//
+// Returns true if a string is a valid IPv4 address, IPv6 address, or is in a CIDR range.
+//
+// - isIPv4: returns true when the IP Address is an IPv4 address.
+//
+// - isIPv6: returns true when the IP Address is an IPv6 address.
+//
+// Examples:
+//
+//	ipAddr('127.0.0.1').isIPv4Addr() // returns true
+//	ipAddr('::1').isIPv6Addr() // returns true
+//	ipAddr('127.0.0.1').isIPv6Addr() // returns false
+//	ipAddr('::1').isIPv4Addr() // returns false
```

Comment:
I would add:

```
isCanonical

Returns true if an address is in canonical form. There is exactly one canonical form
for every IP address, so fields containing IPs in canonical form can just be treated
as strings when checking for equality or uniqueness.

Examples:

     ipAddr('127.0.0.1').isCanonical() // returns true; all valid IPv4 addresses are canonical
     ipAddr('2001:db8::abcd').isCanonical() // returns true
     ipAddr('2001:DB8::ABCD').isCanonical() // returns false
     ipAddr('2001:db8::0:0:0:abcd').isCanonical() // returns false
```

in terms of code, `netip.Addr.String()` always returns canonical form, so you'd just have to add another field to `IPAddr` indicating whether the input string equals `Addr.String()`.

Reply:
oh, and likewise for CIDRs (where a canonical CIDR has a canonical IP address, and a prefix length with no leading 0s)

Reply:
`ServiceCIDR` requires canonical form. :slightly_smiling_face: 

Arguments for requiring canonical form:
- allows code/clients to just treat the fields as strings when checking equality/uniqueness
- helps ensure reconciliation loops don't try to re-reconcile already-correct values

Arguments for not requiring canonical form:
- "I want to directly use the output of some other program that outputs non-canonical IPs without having to fix them up".
- "I DON'T LIKE LOWERCASE LETTERS"

Reply:
As a rule, we do not modify user input. If they gave us something in the API. It's either valid or it's not valid but we don't change it. I don't think you're really suggesting that we would reject IPs in non-conical form though, right?

Reply:
> ServiceCIDR requires canonical form.

Because is an specific implementation and you can specify the input requirements, but I think that in a generic library like this you can not make this assumption

Reply:
> As a rule, we do not modify user input. If they gave us something in the API. It's either valid or it's not valid but we don't change it. I don't think you're really suggesting that we would reject IPs in non-conical form though, right?

Currently `ip` allows non-canonical and users of the CEL library could make their own choice for their own use case, `isIP(<string>) && ip(<string>).isCanonical().

> Require that net == net.Masked(), to limit it to the subnet/mask case rather than the IP-with-prefix-length case.

But for `cidr` @danwinship suggested that we should reject the non-canonical form, but we could add a `masked()` there so users could make their own choice, `isCIDR(<string>) && cidr(<string>) == cidr(<string>).masked()` (I'll have to work out how to make the `==` operator work for the custom library).

I don't have a strong opinion on the networking side of this but can see the API side argument of leaving the library less-opinionated and allowing user choice, so unless @danwinship you think this would be a security issue, are you happy with the library being slightly less opinionated on cidr being canonical form?

Reply:
> As a rule, we do not modify user input.

Right. Wasn't suggesting that.

> I don't think you're really suggesting that we would reject IPs in non-canonical form though, right?

I'm not saying we should change validation of existing APIs to require canonical form, but I think there's a good argument to be made for requiring canonical form in _new_ APIs. Mostly because (a) it has advantages, (b) it doesn't really have disadvantages.

> I think that in a generic library like this you can not make this assumption

Right, that's why I was suggesting it as a separate function; as Joel said, the CRD author can choose whether to require "valid IP" or "valid IP in canonical form".

> But for `cidr` @danwinship suggested that we should reject the non-canonical form

No, I said we should consider IP-with-prefix-length to be not _valid_ for our "CIDR" type. "Canonical" is still a separate consideration. ("`2001:db8::1/64`" is not valid, "`2001:DB8::/64` is valid but not canonical, "`2001:db8::/64`" is valid and canonical).